### PR TITLE
Batch DML support

### LINKED ISSUE #92: Natively support multiple DML statements
Cloud Spanner has several APIs for executing different, although somewhat overlapping, types of queries:
1) [ExecuteStreamingSql](https://cloud.google.com/spanner/docs/reference/rpc/google.spanner.v1#google.spanner.v1.Spanner.ExecuteStreamingSql): supports insert/update statements in addition to the select statements (for which it is most useful). 
However, does NOT support DDL, and requires that DML statements arrive in order specified by `seqno`.
2) [ExecuteBatchDml](https://cloud.google.com/spanner/docs/reference/rpc/google.spanner.v1#google.spanner.v1.Spanner.ExecuteBatchDml): supports multiple insert/update statements.
3) (out of scope for this issue) Database Admin API [UpdateDatabaseDdl](https://cloud.google.com/spanner/docs/reference/rpc/google.spanner.admin.database.v1#google.spanner.admin.database.v1.DatabaseAdmin.UpdateDatabaseDdl). This poses additional complications, as it returns a long-running operation, which then would have to be checked for status separately.

In #87, using multiple `ExecuteStreamingSql` calls when executing parameterized DML queries was made possible by explicitly serializing the DML queries to ensure they arrive to Spanner in order. However, waiting for the network roundtrip for each statement is not ideal. 
Since we will have to detect the nature of each query no matter what implementation we choose (because Spanner has separate APIs for different query types), we should use `ExecuteBatchDml` to execute multiple DML queries.

### PR DESCRIPTION
**Pull request by dmitry-s**
fixes #92

### CODE REVIEW

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java [hunk 1]
```diff
@@ -108,24 +112,27 @@ public Statement bindNull(int i, Class<?> type) {
...
-  private Mono<? extends Result> runSingleStatement(Struct params, StatementType statementType) {
+  private Mono<? extends Result> runSingleStatement(Struct params) {
     PartialResultRowExtractor partialResultRowExtractor = new PartialResultRowExtractor();
 
     Flux<PartialResultSet> resultSetFlux =
         this.client.executeStreamingSql(
             this.session, this.transaction, this.sql, params, this.statementBindings.getTypes());
 
-    if (statementType == StatementType.SELECT) {
+    if (this.statementType == StatementType.SELECT) {
```
Comment (elefeint) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java]: Do we need this check anymore? DML does not go through this method anymore, and I don't imagine DDL will either.

File: src/main/java/com/google/cloud/spanner/r2dbc/client/GrpcClient.java [hunk 2]
```diff
@@ -159,6 +162,31 @@ public GrpcClient(GoogleCredentials credentials) {
...
+    }
+    for (Struct paramsStruct : params) {
+      ExecuteBatchDmlRequest.Statement statement = ExecuteBatchDmlRequest.Statement.newBuilder()
+          .setSql(sql).setParams(paramsStruct).putAllParamTypes(types)
+          .build();
+      request.addStatements(statement);
+    }
+
+    return ObservableReactiveUtil
+        .streamingCall(obs -> this.spanner.executeBatchDml(request.build(), obs));
```
Comment (elefeint) [inline src/main/java/com/google/cloud/spanner/r2dbc/client/GrpcClient.java]: It's not really a streaming call; use `unaryCall()` here.

File: src/test/java/com/google/cloud/spanner/r2dbc/it/SpannerIT.java [hunk 3]
```diff
@@ -268,12 +292,13 @@ public void testEmptySelect() {
   /**
    * Executes a DML query and returns the rows updated.
    */
-  private int executeDmlQuery(String sql) {
+  private List<Integer> executeDmlQuery(String sql) {
```
Comment (dzou) [inline src/test/java/com/google/cloud/spanner/r2dbc/it/SpannerIT.java]: Should this be kept as before, returning integer? B/c `String sql` can only correspond to one statement.

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java [hunk 4]
```diff
@@ -108,24 +112,27 @@ public Statement bindNull(int i, Class<?> type) {
 
   @Override
   public Publisher<? extends Result> execute() {
-    Flux<Struct> structFlux = Flux.fromIterable(this.statementBindings.getBindings());
-    StatementType statementType = StatementParser.getStatementType(this.sql);
-
-    if (statementType == StatementType.SELECT) {
-      return structFlux.flatMap(struct -> runSingleStatement(struct, statementType));
+    if (this.statementType == StatementType.DML) {
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java]: I wonder if you should just have a class hierarchy of statements -- `SpannerDmlStatement`, `SpannerDdlStatement`, `SpannerReadStatement`. WDYT?

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java [hunk 5]
```diff
@@ -108,24 +112,27 @@ public Statement bindNull(int i, Class<?> type) {
...
+          .executeBatchDml(this.session, this.transaction, this.sql,
+              this.statementBindings.getBindings(),
+              this.statementBindings.getTypes())
+          .flatMapIterable(ExecuteBatchDmlResponse::getResultSetsList)
+          .map(resultSet -> new SpannerResult(Flux.empty(),
+              Mono.just(Math.toIntExact(resultSet.getStats().getRowCountExact()))));
     }
-    // DML statements have to be executed sequentially because they need seqNo to be in order
-    return structFlux.concatMapDelayError(struct -> runSingleStatement(struct, statementType));
+    Flux<Struct> structFlux = Flux.fromIterable(this.statementBindings.getBindings());
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java]: Else if `SELECT`? Are you missing the DDL case?
The whole thing cloud be a switch statement, with the default throwing an exception.
Alternatively, see my suggestion about the class hierarchy.

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java (hunk 1 above)
Comment (meltsufin) [reply src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java]: +1 and should probably rename to `runSelectStatement`

File: src/main/java/com/google/cloud/spanner/r2dbc/client/Client.java [hunk 6]
```diff
@@ -83,6 +85,13 @@
     return  executeStreamingSql(session, transaction, sql, null, null);
   }
 
+  /**
+   * Execute DML batch.
+   */
+  Flux<ExecuteBatchDmlResponse> executeBatchDml(Session session,
+      @Nullable SpannerTransactionContext transactionContext, String sql,
+      List<Struct> params, Map<String, Type> types);
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/spanner/r2dbc/client/Client.java]: Should `params` be a `Flux<Struct>`, or is that a higher level concern?

File: src/test/java/com/google/cloud/spanner/r2dbc/it/SpannerIT.java [hunk 7]
```diff
@@ -268,12 +292,13 @@ public void testEmptySelect() {
...
     Connection connection = Mono.from(connectionFactory.create()).block();
 
     Mono.from(connection.beginTransaction()).block();
-    int rowsUpdated = Mono.from(connection.createStatement(sql).execute())
+    List<Integer> rowsUpdated = Flux.from(connection.createStatement(sql).execute())
```
Comment (meltsufin) [inline src/test/java/com/google/cloud/spanner/r2dbc/it/SpannerIT.java]: `s/rowsUpdated/rowsUpdatedPerStatement`

File: src/main/java/com/google/cloud/spanner/r2dbc/client/Client.java (hunk 6 above)
Comment (dmitry-s) [reply src/main/java/com/google/cloud/spanner/r2dbc/client/Client.java]: Yes, it is a higher level concern.

**Pull request merged**